### ISSUE TITLE
Visualize Unreferenced handles

### ISSUE THREAD

**Issue by Fishrock123** [AndreasMadsen/dprof#7]
Not sure exactly how this would be done, as async_wrap doesn't have 'unref' or 'ref' callsites.

Luckily, this is most important for handles that extend to the end of the process.

So I'm thinking we could probably look though the tree of nodes to find ones that were alive at the time of process exit and then check if they were unrefed...

@AndreasMadsen I would probably need a bit of help knowing how to do that, but I'm willing to try given a bit of guidance.

**Comment by AndreasMadsen** [comment]
Sounds awesome. Could you start out by providing an example that highlights the issue, just so we are on the same page?

I will try and update this to use `async_hook`, tomorrow. That should make things a bit easier.

**Comment by Fishrock123** [comment]
@AndreasMadsen Which handle(s) may be keeping the process open?

<img width="1270" alt="screen shot 2016-08-23 at 6 07 36 pm" src="https://cloud.githubusercontent.com/assets/1093990/17911613/8138430a-695c-11e6-9b94-28a11dd3ca9d.png">

In the handle info would be useful, but it would be awesome if we could, say, color the bar green or something rather than blue.

**Comment by AndreasMadsen** [comment]
Great. Lets keep it simple to start with and just assume that the `Unrefed()` value doesn't change over time.

My immediate thought was that we could obtain it at the end of the handles life, either from the `destroy` hook or the `process.on('exit') event. The latter is simple because `stateMap` (in `/dprof.js`) contains a list of all non-destroyed handles. However I'm not sure it is always safe to use the `destroy` hook, as that is sometimes called because the handle was garbage collected. Alternatively, we can maybe monkey patch `.ref()` and `.unref()` in the `init` hook, to get an event like notification.

If you want to try the `destroy` hook solution, be sure to use a `WeekMap` (with `state.add(handle)` as the key), as the handle object will otherwise won't be state.add(handle).

No matter how this `unrefed` flag is obtained, you should:

* add  the flag to the `Node` constructor [/dprof.js#L32](https://github.com/AndreasMadsen/dprof/blob/master/dprof.js#L32).
* export the flag by the `Node.prototype.toJSON` function [/dprof.js#L76](https://github.com/AndreasMadsen/dprof/blob/master/dprof.js#L76).
* the timeline draw logic is in [/visualizer/timeline.js#L179](https://github.com/AndreasMadsen/dprof/blob/master/visualizer/timeline.js#L179).
  * I think you can just add an `unrefed` class to the `g.timeline` node and edit the stylesheet in [/visualizer/index.html](https://github.com/AndreasMadsen/dprof/blob/master/visualizer/index.html).
  * `d3` can add a dynamic class with `.classed({ unrefed: (node) => node.unrefed })`

**Comment by AndreasMadsen** [comment]
It may also be necessary to modify `async-hook` as it doesn't expose `.unrefed()` and friends. You can take a look here: [/async-hook/patches/timers.js#L5](https://github.com/AndreasMadsen/async-hook/blob/master/patches/timers.js#L5)

**Comment by Fishrock123** [comment]
Hmmm this may not be possible without access to the handles at exit...

We'd probably have to use a WeakMap to store references to handles.

Pretty much all unref calls are done asynchronously and there isn't an event for it. Although it is the top on my list once the base AsyncWrap API lands.

Would could possibly do it in `async-hook` as a proof of concept.

**Comment by AndreasMadsen** [comment]
> Edit: not sure how to do that correctly...

Yeah, WeakMaps are fun. What do you think of this: https://github.com/AndreasMadsen/dprof/commit/41240a1cc86f696f23b087ad4a48692112de02c0

It is not exactly perfect, as a unrefed handle could be gc (and then the destroy hook emits), which would cause it to be registred as `refed`.

**Comment by Fishrock123** [comment]
@AndreasMadsen Yeah, it's not a great solution but luckily the handles you really care about here are the ones still alive at process exit.

The [MDN docs on WeakMaps](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/WeakMap) seem to suggest that the "value" is not GC'd so long as the "key" is not. 

This is problematic as IIRC (discussions with @trevnorris) some native handles do not destroy themselves of there are references or something. (?)

**Comment by AndreasMadsen** [comment]
> The MDN docs on WeakMaps seem to suggest that the "value" is not GC'd so long as the "key" is not.

Ah, of cause  :(

Then monkey patching `.unref()` may be the only solution until we have an `ref-changed` event.

**Comment by trevnorris** [comment]
Detecting unref would fall outside the purview of AsyncWrap. Though the idea of adding an event is a potentially good idea, and seems like a good direction.

**Comment by AndreasMadsen** [comment]
> Detecting unref would fall outside the purview of AsyncWrap. Though the idea of adding an event is a potentially good idea, and seems like a good direction.

Agreed. I don't think we can even ensure that `.ref()` and `.unref()` exists on handle objects from userland.

@Fishrock123 I don't think we should wait for AsyncWrap, we should just add some notification system to `node::HandleWrap`. But I'm not sure how that would look (`AddRefListener` `RemoveRefListener` ?).

**Comment by Fishrock123** [comment]
> Agreed. I don't think we can even ensure that .ref() and .unref() exists on handle objects from userland.

I don't really think that should matter. I don't know of any userland APIs that have this functionality, and definitely not as default.

> I don't think we should wait for AsyncWrap, we should just add some notification system to node::HandleWrap. 

Via natives? I mean on `init` we may just be able to wrap `handle.{un}ref()` in JS. (Or in `nextTick` if it is not possible in init ...?

That still doesn't solve the problem that we should be able to keep a reference to a handle until pre-destroy (i.e. `pre`/`post`)

**Comment by AndreasMadsen** [comment]
> I don't really think that should matter. I don't know of any userland APIs that have this functionality, and definitely not as default.
> Keep in mind that if a handle doesn't have this, it is pretty much guaranteed to always be refed.

I agree. But I think a foundation for adding it to AsyncWrap is that it does exists. But it may be sufficient to assume it is refed, I really don't know.

> Via natives? I mean on init we may just be able to wrap handle.{un}ref() in JS. (Or in nextTick if it is not possible in init ...?
> Aside: That still doesn't solve the problem that we should be able to keep a reference to a handle until pre-destroy (i.e. pre/post)

Here is an usage example d70e62e287769090ed71a0f2efb473912012a53b. Because `node` doesn't reference `handle`, nothing should prevent `handle` from being gc'ed.

**Comment by AndreasMadsen** [comment]
#12 landed. There is more to it. But lets create another issue when the time arrives.
